File: sdks/js/src/resilience/retry.ts
/**
 * Retry Policy with Exponential Backoff Implementation
 * 
 * Provides configurable retry logic for transient failures.
 * 
 * @module @aether/sdk/resilience
 */

// ============================================
// Types
// ============================================

export type BackoffStrategy = 'fixed' | 'linear' | 'exponential' | 'exponential-jitter';

export interface RetryConfig {
    /** Maximum number of retry attempts */
    maxAttempts: number;
    /** Backoff strategy */
    backoff: BackoffStrategy;
    /** Base delay in ms */
    baseDelay: number;
    /** Maximum delay in ms */
    maxDelay: number;
    /** Multiplier for exponential backoff */
    multiplier?: number;
    /** Jitter factor (0-1) for jitter strategies */
    jitterFactor?: number;
    /** Predicate to determine if error is retryable */
    isRetryable?: (error: Error, attempt: number) => boolean;
    /** Callback before each retry */
    onRetry?: (error: Error, attempt: number, delay: number) => void;
    /** Callback when all retries exhausted */
    onExhausted?: (error: Error, attempts: number) => void;
}

export interface RetryStats {
    totalAttempts: number;
    successfulAttempts: number;
    failedAttempts: number;
    retriedCalls: number;
    exhaustedCalls: number;
    totalRetryDelayMs: number;
}

export interface RetryResult<T> {
    result: T;
    attempts: number;
    totalDelayMs: number;
}

// ============================================
// Retry Policy Implementation
// ============================================

export class RetryPolicy {
    private stats: RetryStats = {
        totalAttempts: 0,
        successfulAttempts: 0,
        failedAttempts: 0,
        retriedCalls: 0,
        exhaustedCalls: 0,
        totalRetryDelayMs: 0,
    };

    private readonly config: Required<RetryConfig>;

    constructor(config: Partial<RetryConfig> = {}) {
        this.config = {
            maxAttempts: config.maxAttempts ?? 3,
            backoff: config.backoff ?? 'exponential',
            baseDelay: config.baseDelay ?? 100,
            maxDelay: config.maxDelay ?? 30000,
            multiplier: config.multiplier ?? 2,
            jitterFactor: config.jitterFactor ?? 0.1,
            isRetryable: config.isRetryable ?? (() => true),
            onRetry: config.onRetry ?? (() => {}),
            onExhausted: config.onExhausted ?? (() => {}),
        };
    }

    /**
     * Execute a function with retry logic
     */
    async execute<T>(fn: () => Promise<T>): Promise<RetryResult<T>> {
        let attempt = 0;
        let totalDelayMs = 0;
        let lastError: Error | undefined;

        while (attempt < this.config.maxAttempts) {
            attempt++;
            this.stats.totalAttempts++;

            try {
                const result = await fn();
                this.stats.successfulAttempts++;
                
                if (attempt > 1) {
                    this.stats.retriedCalls++;
                }

                return {
                    result,
                    attempts: attempt,
                    totalDelayMs,
                };
            } catch (error) {
                lastError = error as Error;
                this.stats.failedAttempts++;

                // Check if we should retry
                if (attempt >= this.config.maxAttempts || !this.config.isRetryable(lastError, attempt)) {
                    break;
                }

                // Calculate delay
                const delay = this.calculateDelay(attempt);
                totalDelayMs += delay;
                this.stats.totalRetryDelayMs += delay;

                // Notify callback
                this.config.onRetry(lastError, attempt, delay);

                // Wait before retry
                await this.sleep(delay);
            }
        }

        // All retries exhausted
        this.stats.exhaustedCalls++;
        this.config.onExhausted(lastError!, attempt);

        throw new RetryExhaustedError(
            `All ${this.config.maxAttempts} retry attempts exhausted`,
            lastError!,
            attempt,
            totalDelayMs
        );
    }

    /**
     * Execute with a result wrapper (doesn't throw on exhaustion)
     */
    async executeSafe<T>(fn: () => Promise<T>): Promise<RetryResult<T> | null> {
        try {
            return await this.execute(fn);
        } catch (error) {
            if (error instanceof RetryExhaustedError) {
                return null;
            }
            throw error;
        }
    }

    /**
     * Get current statistics
     */
    getStats(): Readonly<RetryStats> {
        return { ...this.stats };
    }

    /**
     * Reset statistics
     */
    resetStats(): void {
        this.stats = {
            totalAttempts: 0,
            successfulAttempts: 0,
            failedAttempts: 0,
            retriedCalls: 0,
            exhaustedCalls: 0,
            totalRetryDelayMs: 0,
        };
    }

    // ============================================
    // Private Methods
    // ============================================

    private calculateDelay(attempt: number): number {
        let delay: number;

        switch (this.config.backoff) {
            case 'fixed':
                delay = this.config.baseDelay;
                break;

            case 'linear':
                delay = this.config.baseDelay * attempt;
                break;

            case 'exponential':
                delay = this.config.baseDelay * Math.pow(this.config.multiplier, attempt - 1);
                break;

            case 'exponential-jitter':
                delay = this.config.baseDelay * Math.pow(this.config.multiplier, attempt - 1);
                delay = this.addJitter(delay);
                break;

            default:
                delay = this.config.baseDelay;
        }

        return Math.min(delay, this.config.maxDelay);
    }

    private addJitter(delay: number): number {
        const jitter = delay * this.config.jitterFactor;
        return delay + (Math.random() * jitter * 2 - jitter);
    }

    private sleep(ms: number): Promise<void> {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

// ============================================
// Retry Error
// ============================================

export class RetryExhaustedError extends Error {
    public readonly lastError: Error;
    public readonly attempts: number;
    public readonly totalDelayMs: number;

    constructor(message: string, lastError: Error, attempts: number, totalDelayMs: number) {
        super(message);
        this.name = 'RetryExhaustedError';
        this.lastError = lastError;
        this.attempts = attempts;
        this.totalDelayMs = totalDelayMs;
    }
}

// ============================================
// Predefined Retry Policies
// ============================================

/**
 * Create a retry policy for transient network errors
 */
export function networkRetryPolicy(overrides: Partial<RetryConfig> = {}): RetryPolicy {
    return new RetryPolicy({
        maxAttempts: 3,
        backoff: 'exponential-jitter',
        baseDelay: 100,
        maxDelay: 5000,
        isRetryable: (error) => {
            // Common transient error patterns
            const transientMessages = [
                'ECONNRESET',
                'ETIMEDOUT',
                'ENOTFOUND',
                'ECONNREFUSED',
                'network',
                'timeout',
                'unavailable',
                'temporary',
            ];
            const message = error.message.toLowerCase();
            return transientMessages.some(m => message.includes(m.toLowerCase()));
        },
        ...overrides,
    });
}

/**
 * Create a retry policy for database operations
 */
export function databaseRetryPolicy(overrides: Partial<RetryConfig> = {}): RetryPolicy {
    return new RetryPolicy({
        maxAttempts: 5,
        backoff: 'exponential',
        baseDelay: 50,
        maxDelay: 2000,
        multiplier: 2,
        isRetryable: (error) => {
            const retryableMessages = [
                'deadlock',
                'lock',
                'busy',
                'timeout',
                'connection',
            ];
            const message = error.message.toLowerCase();
            return retryableMessages.some(m => message.includes(m.toLowerCase()));
        },
        ...overrides,
    });
}

/**
 * Create an aggressive retry policy (many attempts, short delays)
 */
export function aggressiveRetryPolicy(overrides: Partial<RetryConfig> = {}): RetryPolicy {
    return new RetryPolicy({
        maxAttempts: 10,
        backoff: 'exponential-jitter',
        baseDelay: 10,
        maxDelay: 1000,
        multiplier: 1.5,
        jitterFactor: 0.2,
        ...overrides,
    });
}

/**
 * Create a conservative retry policy (few attempts, longer delays)
 */
export function conservativeRetryPolicy(overrides: Partial<RetryConfig> = {}): RetryPolicy {
    return new RetryPolicy({
        maxAttempts: 2,
        backoff: 'exponential',
        baseDelay: 1000,
        maxDelay: 10000,
        multiplier: 3,
        ...overrides,
    });
}

// ============================================
// Decorator for Method Retry
// ============================================

/**
 * Decorator to add retry logic to a method
 */
export function withRetry(policy: RetryPolicy) {
    return function (
        _target: any,
        _propertyKey: string,
        descriptor: TypedPropertyDescriptor<(...args: any[]) => Promise<any>>
    ) {
        const originalMethod = descriptor.value!;

        descriptor.value = async function (...args: any[]) {
            const result = await policy.execute(() => originalMethod.apply(this, args));
            return result.result;
        };

        return descriptor;
    };
}

// ============================================
// Combined Policy Builder
// ============================================

/**
 * Builder for creating combined retry + circuit breaker policies
 */
export class ResiliencePolicyBuilder {
    private retryConfig: Partial<RetryConfig> = {};
    private breakerConfig: Partial<import('./circuit-breaker').CircuitBreakerConfig> = {};

    withRetry(config: Partial<RetryConfig>): this {
        this.retryConfig = { ...this.retryConfig, ...config };
        return this;
    }

    withCircuitBreaker(config: Partial<import('./circuit-breaker').CircuitBreakerConfig>): this {
        this.breakerConfig = { ...this.breakerConfig, ...config };
        return this;
    }

    build(): { retry: RetryPolicy; breaker: import('./circuit-breaker').CircuitBreaker } {
        const { CircuitBreaker } = require('./circuit-breaker');
        
        return {
            retry: new RetryPolicy(this.retryConfig),
            breaker: new CircuitBreaker(this.breakerConfig),
        };
    }

    /**
     * Build a combined executor that uses both policies
     */
    buildExecutor(): <T>(fn: () => Promise<T>) => Promise<T> {
        const { CircuitBreaker } = require('./circuit-breaker');
        
        const retry = new RetryPolicy(this.retryConfig);
        const breaker = new CircuitBreaker(this.breakerConfig);

        return async <T>(fn: () => Promise<T>): Promise<T> => {
            return breaker.execute(async () => {
                const result = await retry.execute(fn);
                return result.result;
            });
        };
    }
}

File: sdks/python/aether_sdk/resilience/retry.py
"""
Retry Policy with Exponential Backoff Implementation

Provides configurable retry logic for transient failures.

Example:
    >>> from aether_sdk.resilience.retry import RetryPolicy, RetryConfig, BackoffStrategy
    >>> policy = RetryPolicy(RetryConfig(max_attempts=5, backoff=BackoffStrategy.EXPONENTIAL))
    >>> result = await policy.execute(my_async_func)
"""

from __future__ import annotations

import asyncio
import random
from dataclasses import dataclass
from enum import Enum
from typing import Awaitable, Callable, Generic, Optional, TypeVar

T = TypeVar("T")


class BackoffStrategy(Enum):
    """Supported backoff strategies for retry delay calculation.

    Attributes:
        FIXED: Constant delay between retries.
        LINEAR: Delay increases linearly with the attempt number.
        EXPONENTIAL: Delay doubles (or multiplies) with each attempt.
        EXPONENTIAL_JITTER: Exponential delay with random jitter to
            avoid thundering-herd effects.
    """

    FIXED = "fixed"
    LINEAR = "linear"
    EXPONENTIAL = "exponential"
    EXPONENTIAL_JITTER = "exponential-jitter"


@dataclass
class RetryConfig:
    """Configuration for a :class:`RetryPolicy`.

    Attributes:
        max_attempts: Maximum number of attempts (including the first).
        backoff: Strategy used to calculate delay between attempts.
        base_delay_ms: Base delay in milliseconds.
        max_delay_ms: Upper bound for the calculated delay.
        multiplier: Multiplier for exponential/linear strategies.
        jitter_factor: Fraction of the delay used for jitter (e.g.
            ``0.1`` means ±10 %).
        is_retryable: Optional callback ``(error, attempt) -> bool``
            that determines whether a specific error should be retried.
        on_retry: Optional callback invoked before each retry.
        on_exhausted: Optional callback invoked when all retries are
            exhausted.
    """

    max_attempts: int = 3
    backoff: BackoffStrategy = BackoffStrategy.EXPONENTIAL_JITTER
    base_delay_ms: int = 100
    max_delay_ms: int = 30000
    multiplier: float = 2.0
    jitter_factor: float = 0.1
    is_retryable: Optional[Callable[[Exception, int], bool]] = None
    on_retry: Optional[Callable[[Exception, int, int], None]] = None
    on_exhausted: Optional[Callable[[Exception, int], None]] = None


@dataclass
class RetryStats:
    """Aggregated statistics for a :class:`RetryPolicy`.

    Attributes:
        total_attempts: Total function invocations.
        successful_attempts: Invocations that returned without error.
        failed_attempts: Invocations that raised an exception.
        retried_calls: Calls that required at least one retry.
        exhausted_calls: Calls that exhausted all retry attempts.
        total_retry_delay_ms: Cumulative delay (ms) spent waiting
            between retries.
    """

    total_attempts: int = 0
    successful_attempts: int = 0
    failed_attempts: int = 0
    retried_calls: int = 0
    exhausted_calls: int = 0
    total_retry_delay_ms: int = 0


@dataclass
class RetryResult(Generic[T]):
    """Result of a successful retry operation.

    Attributes:
        result: The value returned by the async function.
        attempts: Number of attempts required (1 = first try succeeded).
        total_delay_ms: Total time spent waiting between retries.
    """

    result: T
    attempts: int
    total_delay_ms: int


class RetryExhaustedError(Exception):
    """Raised when all retry attempts have been exhausted.

    Attributes:
        last_error: The exception from the final attempt.
        attempts: Total number of attempts made.
        total_delay_ms: Cumulative delay between retries.
    """

    def __init__(
        self, message: str, last_error: Exception, attempts: int, total_delay_ms: int
    ):
        super().__init__(message)
        self.last_error = last_error
        self.attempts = attempts
        self.total_delay_ms = total_delay_ms


class RetryPolicy:
    """Retry policy with configurable backoff strategies.

    Wraps async functions and transparently retries on transient
    failures according to the configured :class:`RetryConfig`.

    Example:
        >>> policy = RetryPolicy()
        >>> result = await policy.execute(fetch_data)
    """

    def __init__(self, config: Optional[RetryConfig] = None):
        """Initialize the retry policy.

        Args:
            config: Optional configuration. Defaults to
                :class:`RetryConfig` with exponential jitter backoff.
        """
        self._config = config or RetryConfig()
        self._stats = RetryStats()

    async def execute(self, func: Callable[[], Awaitable[T]]) -> RetryResult[T]:
        """Execute an async function with retry logic.

        Args:
            func: A zero-argument async callable.

        Returns:
            A :class:`RetryResult` containing the function's return
            value and retry metadata.

        Raises:
            RetryExhaustedError: If all retry attempts fail or the
                error is not retryable.
        """
        attempt = 0
        total_delay_ms = 0
        last_error: Optional[Exception] = None

        while attempt < self._config.max_attempts:
            attempt += 1
            self._stats.total_attempts += 1

            try:
                result = await func()
                self._stats.successful_attempts += 1

                if attempt > 1:
                    self._stats.retried_calls += 1

                return RetryResult(
                    result=result,
                    attempts=attempt,
                    total_delay_ms=total_delay_ms,
                )
            except Exception as error:
                last_error = error
                self._stats.failed_attempts += 1

                is_retryable = (
                    self._config.is_retryable(error, attempt)
                    if self._config.is_retryable
                    else self._is_retryable_default(error)
                )

                if attempt >= self._config.max_attempts or not is_retryable:
                    break

                delay = self._calculate_delay(attempt)
                total_delay_ms += delay
                self._stats.total_retry_delay_ms += delay

                if self._config.on_retry:
                    self._config.on_retry(error, attempt, delay)

                await asyncio.sleep(delay / 1000)

        assert last_error is not None, "last_error should not be None after loop"

        self._stats.exhausted_calls += 1
        if self._config.on_exhausted:
            self._config.on_exhausted(last_error, attempt)

        raise RetryExhaustedError(
            f"All {self._config.max_attempts} retry attempts exhausted",
            last_error,
            attempt,
            total_delay_ms,
        )

    async def execute_safe(
        self, func: Callable[[], Awaitable[T]]
    ) -> Optional[RetryResult[T]]:
        """Execute with retry logic but return ``None`` instead of raising.

        Args:
            func: A zero-argument async callable.

        Returns:
            A :class:`RetryResult` on success, or ``None`` if all
            retries are exhausted.
        """
        try:
            return await self.execute(func)
        except RetryExhaustedError:
            return None

    def get_stats(self) -> RetryStats:
        """Return a snapshot of the retry statistics.

        Returns:
            A copy of the current :class:`RetryStats`.
        """
        return RetryStats(
            total_attempts=self._stats.total_attempts,
            successful_attempts=self._stats.successful_attempts,
            failed_attempts=self._stats.failed_attempts,
            retried_calls=self._stats.retried_calls,
            exhausted_calls=self._stats.exhausted_calls,
            total_retry_delay_ms=self._stats.total_retry_delay_ms,
        )

    def reset_stats(self) -> None:
        """Reset all statistics counters to zero."""
        self._stats = RetryStats()

    def _calculate_delay(self, attempt: int) -> int:
        """Calculate the delay in milliseconds for a given attempt.

        Args:
            attempt: The 1-based attempt number.

        Returns:
            Delay in milliseconds, capped at ``max_delay_ms``.
        """
        delay = 0

        if self._config.backoff == BackoffStrategy.FIXED:
            delay = self._config.base_delay_ms
        elif self._config.backoff == BackoffStrategy.LINEAR:
            delay = self._config.base_delay_ms * attempt
        elif self._config.backoff == BackoffStrategy.EXPONENTIAL:
            delay = self._config.base_delay_ms * (
                self._config.multiplier ** (attempt - 1)
            )
        elif self._config.backoff == BackoffStrategy.EXPONENTIAL_JITTER:
            base = self._config.base_delay_ms * (
                self._config.multiplier ** (attempt - 1)
            )
            delay = self._add_jitter(base)

        return min(int(delay), self._config.max_delay_ms)

    def _add_jitter(self, delay: float) -> int:
        """Add random jitter to a delay value.

        Args:
            delay: The base delay in milliseconds.

        Returns:
            The jittered delay as an integer.
        """
        jitter = delay * self._config.jitter_factor
        return int(delay + random.uniform(-jitter, jitter))

    def _is_retryable_default(self, error: Exception) -> bool:
        """Determine whether an error is transient and retryable.

        Matches common network-related substrings in the error message.

        Args:
            error: The exception to evaluate.

        Returns:
            ``True`` if the error appears to be transient.
        """
        transient_messages = [
            "ECONNRESET",
            "ETIMEDOUT",
            "ENOTFOUND",
            "ECONNREFUSED",
            "network",
            "timeout",
            "unavailable",
            "temporary",
        ]
        message = str(error).lower()
        return any(m.lower() in message for m in transient_messages)


# ============================================
# Predefined Retry Policies
# ============================================


def network_retry_policy(**overrides) -> RetryPolicy:
    """Create a retry policy tuned for transient network errors.

    Uses exponential jitter backoff with sensible defaults for HTTP
    and RPC calls.

    Args:
        **overrides: Keyword arguments forwarded to
            :class:`RetryConfig` (e.g. ``max_attempts=5``).

    Returns:
        A configured :class:`RetryPolicy`.

    Example:
        >>> policy = network_retry_policy(max_attempts=5)
    """
    return RetryPolicy(
        RetryConfig(
            max_attempts=overrides.get("max_attempts", 3),
            backoff=BackoffStrategy.EXPONENTIAL_JITTER,
            base_delay_ms=overrides.get("base_delay_ms", 100),
            max_delay_ms=overrides.get("max_delay_ms", 5000),
        )
    )


def database_retry_policy(**overrides) -> RetryPolicy:
    """Create a retry policy tuned for database operations.

    Uses exponential backoff (no jitter) with short base delays.

    Args:
        **overrides: Keyword arguments forwarded to
            :class:`RetryConfig`.

    Returns:
        A configured :class:`RetryPolicy`.
    """
    return RetryPolicy(
        RetryConfig(
            max_attempts=overrides.get("max_attempts", 5),
            backoff=BackoffStrategy.EXPONENTIAL,
            base_delay_ms=overrides.get("base_delay_ms", 50),
            max_delay_ms=overrides.get("max_delay_ms", 2000),
            multiplier=overrides.get("multiplier", 2.0),
        )
    )


def aggressive_retry_policy(**overrides) -> RetryPolicy:
    """Create an aggressive retry policy with many attempts and short delays.

    Suitable for idempotent operations where rapid recovery is critical.

    Args:
        **overrides: Keyword arguments forwarded to
            :class:`RetryConfig`.

    Returns:
        A configured :class:`RetryPolicy`.
    """
    return RetryPolicy(
        RetryConfig(
            max_attempts=overrides.get("max_attempts", 10),
            backoff=BackoffStrategy.EXPONENTIAL_JITTER,
            base_delay_ms=overrides.get("base_delay_ms", 10),
            max_delay_ms=overrides.get("max_delay_ms", 1000),
            multiplier=overrides.get("multiplier", 1.5),
            jitter_factor=overrides.get("jitter_factor", 0.2),
        )
    )


def conservative_retry_policy(**overrides) -> RetryPolicy:
    """Create a conservative retry policy with few attempts and long delays.

    Suitable for non-idempotent operations or rate-limited APIs.

    Args:
        **overrides: Keyword arguments forwarded to
            :class:`RetryConfig`.

    Returns:
        A configured :class:`RetryPolicy`.
    """
    return RetryPolicy(
        RetryConfig(
            max_attempts=overrides.get("max_attempts", 2),
            backoff=BackoffStrategy.EXPONENTIAL,
            base_delay_ms=overrides.get("base_delay_ms", 1000),
            max_delay_ms=overrides.get("max_delay_ms", 10000),
            multiplier=overrides.get("multiplier", 3.0),
        )
    )
